pkg/etcd/etcd.go

```diff
@@ -853,6 +887,161 @@ func (e *ETCD) Snapshot(ctx context.Context, config *config.Control) error {
 		}
 	}
 
+	snapshots, err := e.listSnapshots(ctx, snapshotDir)
+	if err != nil {
+		return err
+	}
+
+	// store the snapshot data somewhere
+	_ = snapshots
+
+	return nil
+}
+
+// snapshotFile represents a single snapshot and it's
+// metadata.
+type snapshotFile struct {
+	Name      string `json:"name"`
+	NodeName  string `json:"nodeName"`
+	CreatedAt string `json:"createdAt"`
+	Size      int64  `json:"size"`
+}
+
+// listSnapshots provides a list of the currently stored
+// snapshots on disk or in S3 along with their relevant
+// metadata.
+func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]snapshotFile, error) {
+	var snapshots []snapshotFile
+
+	if e.config.EtcdS3 {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		objects := e.s3.client.ListObjects(ctx, e.config.EtcdS3BucketName, minio.ListObjectsOptions{
+			Prefix: e.s3.snapshotPrefix(),
+		})
+
+		for obj := range objects {
+			if obj.Err != nil {
+				return nil, obj.Err
+			}
+			snapshots = append(snapshots, snapshotFile{
+				Name:      "s3://" + e.config.EtcdS3BucketName + "/" + obj.Key,
+				CreatedAt: obj.LastModified.String(),
+				Size:      obj.Size,
+			})
+		}
+
+		return snapshots, nil
+	}
+
+	files, err := ioutil.ReadDir(snapshotDir)
+	if err != nil {
+		return nil, err
+	}
+
+	nodeName := os.Getenv("NODE_NAME")
+
+	for _, f := range files {
+		snapshots = append(snapshots, snapshotFile{
+			Name:      filepath.Join(snapshotDir, f.Name()),
+			NodeName:  nodeName,
+			CreatedAt: f.ModTime().String(),
+			Size:      f.Size(),
+		})
+	}
+
+	return snapshots, nil
+}
+
+// updateConfigMap
+func updateConfigMap(data map[string]string, snapshotFiles []snapshotFile) error {
+	for _, v := range snapshotFiles {
+		b, err := json.Marshal(v)
+		if err != nil {
+			return err
+		}
+		data[v.Name] = string(b)
+	}
+
+	return nil
+}
+
+// storeSnapshotData stores the given snapshot data in the "snapshots" ConfigMap.
+func (e *ETCD) storeSnapshotData(ctx context.Context, snapshotFiles []snapshotFile) error {
+	snapshotConfigMap, err := e.cs.CoreV1().ConfigMaps(metav1.NamespaceSystem).Get(ctx, "snapshots", metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			data := make(map[string]string, len(snapshotFiles))
+			if err := updateConfigMap(data, snapshotFiles); err != nil {
+				return err
+			}
+			cm := v1.ConfigMap{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "ConfigMap",
+					APIVersion: "v1",
+				},
```

Comment:
This shouldn't be necessary, the controller should fill this in for you when you ship it to the apiserver